ceno_zkvm/src/structs.rs

```diff
@@ -460,20 +460,20 @@ impl<E: ExtensionField> ZKVMWitnesses<E> {
         cs: &ZKVMConstraintSystem<E>,
         (shard_ctx, final_mem): &(
             &ShardContext,
-            &[(&'static str, InstancePaddingStrategy, &[MemFinalRecord])],
+            &[(&'static str, Option<Range<Addr>>, &[MemFinalRecord])],
         ),
         config: &<ShardRamCircuit<E> as TableCircuit<E>>::TableConfig,
     ) -> Result<(), ZKVMError> {
         let perm = <E::BaseField as PoseidonField>::get_default_perm();
-        let waddr_first_access = if shard_ctx.is_first_shard() {
-            shard_ctx.get_addr_accessed_first_shard()
-        } else {
-            FxHashSet::default()
-        };
+        let addr_accessed = shard_ctx.get_addr_accessed();
 
-        let non_first_shard_records = if shard_ctx.is_first_shard() {
+        // 1. process final mem which ONLY init in first shard
```

Comment:
These are memory records that are **ONLY** initialized in 0th shard but are also accessed in future, right?